### PR TITLE
feat: encode fastq downloader

### PR DESCRIPTION
**Pull request by Maarten-vd-Sande**
### Description

Download fastq files directly from the ENCODE project. 

Added python as a dependency because wrapper **needs** an environment: https://github.com/snakemake/snakemake/issues/1718

### QC

* [x] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* the `meta.yaml` contains a link to the documentation of the respective tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with [black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as conda-forge should have highest priority and defaults channels are usually not needed because most packages are in conda-forge nowadays).

### CODE REVIEW

**Comment by Maarten-vd-Sande** [comment]
@dlaehnemann could you take a look if this is what you had in mind for the seq2science review?

File: bio/encode_fastq_downloader/test/Snakefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+rule encode_fastq_download_PE:
+    output:
+        "{accession}_R1.fastq.gz",
+        "{accession}_R2.fastq.gz"
```
Comment (dlaehnemann) [inline bio/encode_fastq_downloader/test/Snakefile:4]: I think you could work with named outputs here, to avoid relying on the order of specified outputs. Here, I would suggest:

```suggestion
        r1="{accession}_R1.fastq.gz",
        r2="{accession}_R2.fastq.gz"
```

Below, for the single-end case, I would then suggest something like `se=` or `r0=`. Then, you can check for the presence of `r1` and `r2`, or `se`/`r0` in `snakemake.output`.

File: bio/encode_fastq_downloader/environment.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+channels:
+  - conda-forge
+  - bioconda
+  - nodefaults
+dependencies:
+  - snakemake-wrapper-utils >=0.6.2
```
Comment (dlaehnemann) [inline bio/encode_fastq_downloader/environment.yaml:6]: You don't import (let alone use) this in the wrapper.py script, right? So you don't actually need an environment.yaml, I guess?

File: bio/encode_fastq_downloader/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        exception_to_log(
+            check=inferred_layout == "single-ended",
+            msg=f"""The sample was automatically inferred to be single-ended, but it is: "{inferred_layout}".""",
+        )
+        for encff_accession in response["@graph"]:
+            encff_accession = encff_accession["accession"]
+            download_encff(encff_accession, layout, dest)
+    elif layout == "paired":
+        exception_to_log(
+            check=inferred_layout == "single-ended",
```
Comment (dlaehnemann) [inline bio/encode_fastq_downloader/wrapper.py:94]: Copy-pasta, I guess:
```suggestion
            check=inferred_layout == "paired-ended",
```

**Review by dlaehnemann — COMMENTED** [review]
Yes, this is exactly what I was thinking of. This is really great, many thanks!

I already have some small suggestions, but will have a more thorough tomorrow.

File: bio/encode_fastq_downloader/environment.yaml (hunk 2 above)
Comment (Maarten-vd-Sande) [reply bio/encode_fastq_downloader/environment.yaml]: I somehow got the impression the docs woudln't build without the environment.yaml, so I just put something in there for the time being... :upside_down_face: 

I removed the env file, and the docs seem to build

Comment (Maarten-vd-Sande) [reply bio/encode_fastq_downloader/environment.yaml]: Nvm, `Tests / docs` passes, but `Tests / testing` doesn't :upside_down_face:

Comment (dlaehnemann) [reply bio/encode_fastq_downloader/environment.yaml]: I think you have successfully identified a corner case, here... :sweat_smile: 

If I understand the `snakemake` code correctly, it will **always** look for an `environment.yaml`:

1. If the rule is a wrapper, it goes looking with `get_conda_env`:
https://github.com/snakemake/snakemake/blob/27b224ed12448df8aebc7d1ff8f25e3bf7622232/snakemake/workflow.py#L1678
2. `get_conda_env()` doesn't do any checking for existence:
https://github.com/snakemake/snakemake/blob/27b224ed12448df8aebc7d1ff8f25e3bf7622232/snakemake/wrapper.py#L62-L67

So I think you are right that the wrapper needs an `environment.yaml`. But we could create one without any packages specified; and even without any channels, I'd say. Can we try that?

Comment (Maarten-vd-Sande) [reply bio/encode_fastq_downloader/environment.yaml]: Empty environment.yaml also doesn't work

Comment (Maarten-vd-Sande) [reply bio/encode_fastq_downloader/environment.yaml]: I decided to specify a minimum python version as this is usually the case anyways

File: bio/encode_fastq_downloader/environment.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+channels:
+  - conda-forge
+  - bioconda
+  - nodefaults
+# there are strictly no dependencies for this wrapper, but
+# the tests require a environment.yaml with 
+# dependencies. So this is just to please the tests :)
```
Comment (dlaehnemann) [inline bio/encode_fastq_downloader/environment.yaml:7]: I would explicitly mention the issue over at the snakemake repo, so we can find it again in the future:
```suggestion
# There are strictly no dependencies for this wrapper, but the current handling
# of wrappers by snakemake requires an environment.yaml with dependencies.
# Once the underlying issue is fixed, this environment.yaml can be removed. See:
# https://github.com/snakemake/snakemake/issues/1718
```

We should also mention this issue either in the commit message of this change, or somewhere in the pull request. That way we get cross-references on GitHub.

File: bio/encode_fastq_downloader/test/Snakefile (hunk 1 above)
Comment (Maarten-vd-Sande) [reply bio/encode_fastq_downloader/test/Snakefile]: I'm not sure what to call the single-end named outputs. Based on Illumina's explanation I'm thinking to just name them r1, and r1 + r2 for paired-end: https://knowledge.illumina.com/software/general/software-general-reference_material-list/000002211

> For a single-read run, one Read 1 (R1) FASTQ file is created for each sample per flow cell lane. For a paired-end run, one R1 and one Read 2 (R2) FASTQ file is created for each sample for each lane. FASTQ files are compressed and created with the extension *.fastq.gz.

Not entirely convinced of this, but perhaps better than thinking of our own format?

Comment (dlaehnemann) [reply bio/encode_fastq_downloader/test/Snakefile]: You are right, from the general logic of Illumina, single-end reads simply sequence `read1`. `se` and `r0` are just things I have seen in the wild and never questioned. But it sounds better and more consistent to go for your suggestion:
* `r1=` only for single-end
* `r1=` and `r2=` for paired-end

And you can still clearly introduce logic for handling both cases (and throwing an error if neither is the case). Thanks for being so thorough!

**Review by johanneskoester — APPROVED** [review]
